src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,668 @@
+}
+
+/**
+ * We have to distinguish INVALID value with UNSUPPORTED value.
+ * INVALID means the value is invalid in Spark SQL.
+ * UNSUPPORTED means the value is valid in Spark SQL but not supported by rapids
+ * yet. As for INVALID values, we treat them in the same as Spark SQL. As for
+ * UNSUPPORTED values, we just throw cuDF exception.
+ */
+enum ParseResult { OK = 0, INVALID = 1, UNSUPPORTED = 2 };
+
+template <bool with_timezone>
+struct parse_timestamp_string_fn {
+  column_device_view const d_strings;
+  size_type default_tz_index;
+  bool allow_tz_in_date_str = true;
+  // The list column of transitions to figure out the correct offset
+  // to adjust the timestamp. The type of the values in this column is
+  // LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32,
+  // looseTzInstant: int64>>.
+  thrust::optional<lists_column_device_view const> transitions = thrust::nullopt;
+  thrust::optional<column_device_view const> tz_indices        = thrust::nullopt;
+  thrust::optional<column_device_view const> tz_short_ids      = thrust::nullopt;
+
+  __device__ thrust::tuple<cudf::timestamp_us, uint8_t> operator()(const cudf::size_type& idx) const
+  {
+    // inherit the null mask of the input column
+    if (!d_strings.is_valid(idx)) {
+      return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}}, ParseResult::INVALID);
+    }
+
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+
+    timestamp_components ts_comp{};
+    char const* tz_lit_ptr = nullptr;
+    size_type tz_lit_len   = 0;
+    switch (parse_string_to_timestamp_us(&ts_comp, &tz_lit_ptr, &tz_lit_len, d_str)) {
+      case ParseResult::INVALID:
+        return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}}, ParseResult::INVALID);
+      case ParseResult::UNSUPPORTED:
+        return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}},
+                                  ParseResult::UNSUPPORTED);
+      case ParseResult::OK:
+      default: break;
+    }
+
+    if constexpr (!with_timezone) {
+      // path without timezone, in which unix_timestamp is straightforwardly
+      // computed
+      auto const ts_unaligned = compute_epoch_us(ts_comp);
+      return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{ts_unaligned}},
+                                ParseResult::OK);
+    }
+
+    // path with timezone, in which timezone offset has to be determined before
+    // computing unix_timestamp
+    int64_t tz_offset;
+    if (tz_lit_ptr == nullptr) {
+      tz_offset = extract_timezone_offset(compute_loose_epoch_s(ts_comp), default_tz_index);
+    } else {
+      auto tz_view = string_view(tz_lit_ptr, tz_lit_len);
+
+      // Map short TZ ID to region-based timezone if tz_view is a short ID
+      auto const& short_name_col   = tz_short_ids->child(0);
+      auto const& region_based_col = tz_short_ids->child(1);
+      for (size_type i = 0; i < tz_short_ids->size(); i++) {
```

Comment:
Can we please sort the IDs and do a binary search?

Reply:
Done.